### PR TITLE
Refactor Tabs and Table-Sources

### PR DESCRIPTION
**Pull request by deiferni**
This PR attempts a cleanup of `opengever.tabbedview` for our overridden`ftw.tabbedview` and `ftw.table`code.

Goals/Motivation:
- provide better base classes for tabs, you can now inherit from `BaseListingTab` for listings instead of `grok.View, OpengeverTab, ListingView`
- make base classes importable to the `opengever.tabbedview` package directly instead of from some submodule
- Provide support for state filters in all tabs with minimal effort
- remove duplicate code (there were several copies of `ITableSourceConfig.build_query`
- drop dead code
- drop clients-listing, it has been broken for more than one year. obviously we don't care 😉 .

### CODE REVIEW

File: opengever/tabbedview/__init__.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+from opengever.tabbedview.basesource import GeverTableSource
```
Comment (lukasgraf) [inline opengever/tabbedview/__init__.py]: This is a convenience import that would be complained about by Flake8, right? Could you maybe suffix this with `# noqa` so it doesn't get removed?

File: opengever/tabbedview/browser/ogdscontrolpanel.py [hunk 2]
```diff
@@ -20,11 +20,6 @@ class OGDSControlPanel(grok.View, TabbedView):
     grok.require('cmf.ManagePortal')
 
     tabs = [
-        {'id': 'ogds-cp-clients',
-         'icon': None,
-         'url': '#',
-         'class': None},
-
         {'id': 'users',
```
Comment (phgross) [inline opengever/tabbedview/browser/ogdscontrolpanel.py]: @deiferni @lukasgraf Brauchts die Benutzerauflistung im Controlpanel noch? Aus meiner Sicht kann man die entfernen, ist ja unter Kontakte ersichtlich ...

File: opengever/tabbedview/browser/tasklisting.py [hunk 3]
```diff
@@ -140,39 +134,18 @@ def build_query(self):
...
 
-        # get the base query from the config
-        query = self.config.get_base_query()
-        query = self.validate_base_query(query)
+    def avoid_duplicates(self, query):
+        """If a task has a successor task, list only one of them.
 
-        # If a task has a successor task, list only one of them.
-        # List the only the one which is assigned to this client.
+        List the only the one which is assigned to this client.
```
Comment (phgross) [inline opengever/tabbedview/browser/tasklisting.py]: `List the only the one ` --> `List the only one` ?

File: opengever/tabbedview/__init__.py [hunk 4]
```diff
@@ -1,3 +1,9 @@
+from opengever.tabbedview.base_tabs import BaseCatalogListingTab
+from opengever.tabbedview.base_tabs import BaseListingTab
+from opengever.tabbedview.base_tabs import GeverTabMixin
+from opengever.tabbedview.basesource import GeverTableSource
+from opengever.tabbedview.catalog_source import GeverCatalogTableSource
+from opengever.tabbedview.sqlsource import SqlTableSource
```
Comment (lukasgraf) [inline opengever/tabbedview/__init__.py]: These are convenience import that would be complained about by Flake8, right? Could you maybe suffix them with `# noqa` so it doesn't get removed, or a comment that prevents a future, over-eager developer from removing them?

File: opengever/contact/contactfolder.py [hunk 5]
```diff
@@ -101,7 +101,7 @@ class Contacts(OpengeverCatalogListingTab):
     selection = ViewPageTemplateFile("no_selection_amount.pt")
 
     def update_config(self):
-        OpengeverCatalogListingTab.update_config(self)
+        BaseCatalogListingTab.update_config(self)
```
Comment (phgross) [inline opengever/contact/contactfolder.py]: Das sollte nun mit einem super(Contacts, self).update_config() funktionieren oder ?

File: opengever/tabbedview/browser/ogdscontrolpanel.py (hunk 2 above)
Comment (deiferni) [inline opengever/tabbedview/browser/ogdscontrolpanel.py]: also es funktioniert im moment noch 😉 , falls ausbauen erwünscht ware das hier out of scope.

*(vermutlich könne man das ganze controlpanel entfernen, alle aufgaben sieht man ja auch in der persönlichen übersicht.)*

Comment (lukasgraf) [inline opengever/tabbedview/browser/ogdscontrolpanel.py]: Brauchts IMHO nicht mehr, würde das aber auch in einem separaten PR droppen.

**Comment by phgross** [comment]
@deiferni wäre es evtl. sinnvoll zuerst den Bug #1850 zu analysieren bzw. zu korrigieren bevor wir den hier mergen?

File: opengever/tabbedview/browser/tasklisting.py (hunk 3 above)
Comment (lukasgraf) [inline opengever/tabbedview/browser/tasklisting.py]: `List only the one`

**Comment by lukasgraf** [comment]
Looks good to me overall! 👍 💚 

I'll just quickly look into the MRO trickery and circular import issues some more, to see whether we can't find a cleaner solution.

**Comment by deiferni** [comment]
i fear there is a fundamental design problem with the commands stuff since circular import issues seem to accumulate there 😞 , can't put my finger on it yet, though.

afaik the `MRO` trickery is necessary at the moment since we use `grok` and there is some reason that forces us to inherit from `grok.View` as the first baseclass

**Comment by deiferni** [comment]
@phgross zu deiner Bemerkung wegen https://github.com/4teamwork/opengever.core/issues/1850, ich habs analysiert und das problem ist für diesen PR irrelevant.

File: opengever/tabbedview/sqlsource.py [hunk 6]
```diff
@@ -1,23 +1,26 @@
...
 from sqlalchemy.orm.query import Query
-from sqlalchemy.sql.expression import asc, desc
+from sqlalchemy.sql.expression import asc
 from sqlalchemy.sql.expression import column
+from sqlalchemy.sql.expression import desc
 from zope.interface import Interface
 
 
-class SqlTableSource(grok.MultiAdapter, BaseTableSource):
+class SqlTableSource(grok.MultiAdapter, FilteredTableSourceMixin, BaseTableSource):
```
Comment (lukasgraf) [inline opengever/tabbedview/sqlsource.py]: As discussed: Should also inherit from `basesource.GeverTableSource`.

**Comment by deiferni** [comment]
@lukasgraf aktualisiert

**Comment by lukasgraf** [comment]
👍

**Pull request merged**